alg/viewshed/cumulative_viewshed.cpp

```diff
@@ -0,0 +1,196 @@
+/******************************************************************************
+ *
+ * Project:  Viewshed Generation
+ * Purpose:  Core algorithm implementation for viewshed generation.
+ *
+ ******************************************************************************
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+const int NUM_JOBS = 2;
+
+#include <algorithm>
+#include <limits>
+#include <thread>
+
+#include "cpl_worker_thread_pool.h"
+
+#include "combiner.h"
+#include "cumulative_viewshed.h"
+#include "notifyqueue.h"
+#include "util.h"
+#include "viewshed_executor.h"
+
+namespace gdal
+{
+namespace viewshed
+{
+
+Cumulative::Cumulative(const Options &opts) : m_opts(opts)
+{
+}
+
+/// Compute the cumulative viewshed of a raster band.
+///
+/// @param pfnProgress  Pointer to the progress function. Can be null.
+/// @param pProgressArg  Argument passed to the progress function
+/// @return  True on success, false otherwise.
+bool Cumulative::run(const std::string &srcFilename,
+                     GDALProgressFunc pfnProgress, void *pProgressArg)
+{
+    // In cumulative view, we run the executors in normal mode and want "1" where things
+    // are visible.
+    m_opts.outputMode = OutputMode::Normal;
+    m_opts.visibleVal = 1;
+
+    //ABELL
+    /**
+    if (!setupProgress(pfnProgress, pProgressArg))
+        return false;
+    **/
+    DatasetPtr srcDS(
```

Comment:
srcDS should be checked for null

Reply:
More importantly, I forgot to turn the progress stuff back on ;)

Reply:
Fixed.